### ISSUE TITLE
World Position to Image Coordinates (Hololens 2)

### ISSUE THREAD

**Issue by felderto** [MicrosoftDocs/mixed-reality#343]
Hello everyone

I am currently working with Vuforia on Hololens 2 in Unity 2019.4.28f1. I track 4 points (corners) of a Marker and then take an Image of that scene (PhotoCapture class), while also saving the World Space Coordinates of the 4 points that I track, and also the cameraToWorldMatrix and projectionMatrix with the help of the according methods in the PhotoCaptureFrame class from Unity.Windows.WebCam namespace.

My goal is to convert these 4 points from World Space to Pixel Coordinates so that I can later locate these points on the image that I took with the PhotoCaptureFrame Class.

Unfortunately, the documentation for the [TryGetProjectionMatrix Method](https://docs.unity3d.com/ScriptReference/Windows.WebCam.PhotoCaptureFrame.TryGetProjectionMatrix.html?_gl=1*gqmwtz*_ga*MTY0NDg2NjQ3Ny4xNjIzMjI0MTcx*_ga_1S78EFL1W5*MTYyMzM5OTQwOC43LjEuMTYyMzM5OTQ5NC41OQ..&_ga=2.178194535.318410070.1623224171-1644866477.1623224171) is rather sparse.

Everything seemed to work at first, since I receive some cameraToWorldMatrix and projectionMatrix from the PhotoCaptureFrame Class. I save the image by exporting the texture to a PNG format with the EncodeToPNG() Method.

Now my Question:

How is the projectionMatrix composed? Are camera intrinsics included? It is not clear from the documentation.

I tried to get the Pixel coordinates in the following way (source is [this work](https://uwspace.uwaterloo.ca/bitstream/handle/10012/12979/PoothicottuJacob_George.pdf?sequence=1&isAllowed=y)):

`Vector3 CameraSpacePos = WorldToCam.MultiplyPoint(WorldSpacePos);`
`Vector3 ImagePosUnnormalized = ProjMat.MultiplyPoint(CameraSpacePos);`
`Vector2 ImagePosProjected = new Vector2(ImagePosUnnormalized.x, ImagePosUnnormalized.y) / ImagePosUnnormalized.z;`
`Vector2 ImagePosZeroToOne = (ImagePosProjected * 0.5f) + new Vector2(0.5f, 0.5f);`
`Vector2 PixelPos = new Vector2(ImagePosZeroToOne.x * Tex.width, (1 - ImagePosZeroToOne.y) * Tex.height);`

But this gives me the wrong image coordinates. The projection is on the image plane but it seems like some scale/translation factor is wrong. What am I doing wrong in the calculations? Can someone tell me how to calculate the exact pixel coordinates from the 3D world position?

[Here](https://drive.google.com/file/d/1VsJWB2LX_J1lIWEoLBmUS0y9E5nY04Gv/view?usp=sharing) is an image of the wrongly calculated corner points.

Thanks for your help guys!

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 2dd07cb8-0169-9eeb-706a-890cd0418d0e
* Version Independent ID: 46b95d28-2e43-1c9d-a338-495210bd9f91
* Content: [Photo Video camera in Unity - Mixed Reality](https://docs.microsoft.com/en-us/windows/mixed-reality/develop/unity/locatable-camera-in-unity)
* Content Source: [mixed-reality-docs/mr-dev-docs/develop/unity/locatable-camera-in-unity.md](https://github.com/MicrosoftDocs/mixed-reality/blob/docs/mixed-reality-docs/mr-dev-docs/develop/unity/locatable-camera-in-unity.md)
* Product: **mixed-reality**
* Technology: **mr-dev**
* GitHub Login: @keveleigh
* Microsoft Alias: **v-hferrone**

**Comment by dfields-msft** [comment]
Yes, the projection matrix includes the camera intrinsics. @keveleigh may be able to provide some additional details, but this documentation should be up to date and accurate (if not terribly clear/concise); see specifically the sample code in the linked Unity forum thread:
https://docs.microsoft.com/en-us/windows/mixed-reality/develop/platform-capabilities-and-apis/locatable-camera#using-unity
https://docs.microsoft.com/en-us/windows/mixed-reality/develop/unity/locatable-camera-in-unity#locatable-camera
https://forum.unity.com/threads/holographic-photo-blending-with-photocapture.416023/?_ga=2.57872105.210548785.1614215615-862490274.1597860099

The TL;DR should be that the projection matrix transforms 3D coordinates in the camera's view space (i.e. in meters) to **homogeneous** normalized display coordinates, with the image's principal point centered on the Z axis of the camera's view space. Critically, the X and Y components must be divided by W to obtain non-homogeneous image plane coordinates in the -1,1 range. This seems to be the issue with the code you posted; you are dividing by Z instead of W. Do you get better results if you divide by W instead? (Note line 60 of the HolographicImageBlendShader.shader code in the Unity forum post linked above, which does exactly this.)

**Comment by felderto** [comment]
Thanks for the answer. It finally worked out, also with the help of your hint (dividing by W instead of Z). 

For the post processing part I switched over to matlab which made it a lot easier to debug the whole code. 

Thanks again !

![98769590-05A8-4A74-B64E-0A51C1ACFA6D](https://user-images.githubusercontent.com/34155982/123055228-f5481900-d405-11eb-8cec-e6aba35219ef.jpeg)

**Comment by flashchen73** [comment]
> Thanks for the answer. It finally worked out, also with the help of your hint (dividing by W instead of Z).
> 
> For the post processing part I switched over to matlab which made it a lot easier to debug the whole code.
> 
> Thanks again !
> 
> ![98769590-05A8-4A74-B64E-0A51C1ACFA6D](https://user-images.githubusercontent.com/34155982/123055228-f5481900-d405-11eb-8cec-e6aba35219ef.jpeg)

Hello, I still don't quite understand how you solved this. Where did this W come from? Is it Vector4? I have tried using Vector4. W, but this value is 0, and the final result is inf representing infinity. Can you tell me how you solved it?
